agent/agent.go

```diff
@@ -889,23 +912,22 @@ func (a *agent) handleReconnectingPTY(ctx context.Context, msg codersdk.Reconnec
 			rpty.Close()
 			a.reconnectingPTYs.Delete(msg.ID)
 		}); err != nil {
-			a.logger.Error(ctx, "start reconnecting pty routine", slog.F("id", msg.ID), slog.Error(err))
-			return
+			return xerrors.Errorf("start routine: %w", err)
 		}
 	}
 	// Resize the PTY to initial height + width.
 	err := rpty.ptty.Resize(msg.Height, msg.Width)
 	if err != nil {
 		// We can continue after this, it's not fatal!
-		a.logger.Error(ctx, "resize reconnecting pty", slog.F("id", msg.ID), slog.Error(err))
+		logger.Error(ctx, "resize", slog.Error(err))
 	}
 	// Write any previously stored data for the TTY.
 	rpty.circularBufferMutex.RLock()
-	_, err = conn.Write(rpty.circularBuffer.Bytes())
+	prevBuf := slices.Clone(rpty.circularBuffer.Bytes())
 	rpty.circularBufferMutex.RUnlock()
+	_, err = conn.Write(prevBuf)
```

Comment:
Review: Unrelated fix(?). It seemed like a bad idea (to me) to perform a write on a network connection whilst holding the mutex, this shortens the span.

However, neither now, nor before, do we achieve perfect synchronization. In theory we should ensure that writes happening between writing `prevBuf` to conn and adding conn to `activeConns` are not lost. I did not try to fix this issue in this PR.

Reply:
Makes sense. It might be worth leaving that in a comment.